sqlanalytics/resource_query.go

```diff
@@ -0,0 +1,345 @@
+}
+
+func (r *queryResource) fromAPIObject(aq *api.Query, d *schema.ResourceData) error {
+	var q Query
+
+	// Transform from API object.
+	q.ID = aq.ID
+	q.DataSourceID = aq.DataSourceID
+	q.Name = aq.Name
+	q.Description = aq.Description
+	q.Query = aq.Query
+
+	if s := aq.Schedule; s != nil {
+		q.Schedule = &QuerySchedule{
+			Interval: s.Interval,
+		}
+	}
+
+	if len(aq.Tags) > 0 {
+		q.Tags = append(q.Tags, aq.Tags...)
+	}
+
+	if aq.Options != nil {
+		for _, ap := range aq.Options.Parameters {
+			var p QueryParameter
+			switch apv := ap.(type) {
+			case *api.QueryParameterText:
+				p.Name = apv.Name
+				p.Title = apv.Title
+				p.Type = "text"
+				p.Value = apv.Value
+			case *api.QueryParameterNumber:
+				p.Name = apv.Name
+				p.Title = apv.Title
+				p.Type = "number"
+				p.Value = apv.Value
+			case *api.QueryParameterEnum:
+				p.Name = apv.Name
+				p.Title = apv.Title
+				p.Value = apv.Value
+				p.Enum = &QueryParameterEnum{
+					Options:  strings.Split(apv.Options, "\n"),
+					Multiple: newQueryParameterAllowMultiple(apv.Multi),
+				}
+			case *api.QueryParameterQuery:
+				p.Name = apv.Name
+				p.Title = apv.Title
+				p.Value = apv.Value
+				p.Query = &QueryParameterQuery{
+					QueryID:  apv.QueryID,
+					Multiple: newQueryParameterAllowMultiple(apv.Multi),
+				}
+			default:
+				log.Fatalf("Don't know what to do for type: %#v", reflect.TypeOf(apv).String())
+			}
+
+			q.Parameter = append(q.Parameter, p)
+		}
+	}
+
+	// Transform to ResourceData.
+	if err := common.StructToData(q, r.schema, d); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *queryResource) create(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+	aq, err := r.toAPIObject(d)
+	if err != nil {
+		return err
+	}
+
+	var w = api.NewWrapper(ctx, c)
+	aqNew, err := w.CreateQuery(aq)
+	if err != nil {
+		return err
+	}
+
+	err = r.fromAPIObject(aqNew, d)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(aqNew.ID)
+	return nil
+}
+
+func (r *queryResource) read(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+	aq, err := r.toAPIObject(d)
+	if err != nil {
+		return err
+	}
+
+	var w = api.NewWrapper(ctx, c)
+	aqNew, err := w.ReadQuery(aq)
+	if err != nil {
+		return err
+	}
+
+	err = r.fromAPIObject(aqNew, d)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *queryResource) update(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+	aq, err := r.toAPIObject(d)
+	if err != nil {
+		return err
+	}
+
+	var w = api.NewWrapper(ctx, c)
+	aqNew, err := w.UpdateQuery(aq)
+	if err != nil {
+		return err
+	}
+
+	err = r.fromAPIObject(aqNew, d)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *queryResource) delete(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+	aq, err := r.toAPIObject(d)
+	if err != nil {
+		return err
+	}
+
+	var w = api.NewWrapper(ctx, c)
+	err = w.DeleteQuery(aq)
```

Comment:
refactor it to just `return api.NewWrapper(ctx, c).DeleteQuery(d.Id())`, don't make this method longer than it has to be :-)

Reply:
Done.